### ISSUE TITLE
Advice on how to write custom renderers for Vega

### ISSUE THREAD

**Issue by nsonnad** [vega/vega#795]
Hello, I am looking to write a set of custom renderers for Vega, but need some help on how best to proceed. I would like to use Vega right up to the point of actually drawing a chart, including using Vega's JSON description of a chart and the code that parses this JSON representation.

The reasons for this are that 1) the Vega JSON language is a very flexible and effective way to describe and store a chart, and 2) you guys have already figured out how to parse this JSON description, so it seems silly to reimplement parsing. However, my use case has some different opinions about how to draw charts that cannot be accounted for just by changing the config settings of the Vega renderer. Therefore, I'd like to just use Vega's JSON and parsing of it, then insert my own rendering code.

Does this seem like a realizable goal?

It seems so to me in theory, but I have looked into the various vega libraries and am not sure at what point I would be able to get a parsed version of all of a chart's settings (based on its Vega JSON description) and insert my rendering code. The output of `vega-runtime` seems pretty close—it has the `data` field with relevant transforms applied (eg `stack`), and has processed all scales into d3 scales based on the `domain` settings and so on. But it looks like, for `vega-runtime` to work, I need `vega-view`, which seems to need `vega-scenegraph`, which is just vega's renderer.

TL;DR of my question is: Is it possible to just use a few of vega's libraries to parse a Vega JSON spec and get just the transformed, parsed settings that I can plug into another rendering library? If so, which libraries can I use to achieve that?

Thanks (and thanks to Arvind who has already discussed this with me a bit over email)

**Comment by jheer** [comment]
Vega is designed to allow extensible _renderers_, but requires that one use Vega's generated _scenegraph_. The job of a renderer is thus to take a scenegraph (and potentially container DOM element) as input, and draw that scenegraph (or a "damaged" subset thereof).

The best place to start is the [vega-scenegraph](https://github.com/vega/vega-scenegraph) repository, which provides three different renderers: Canvas, SVG and SVGString. The latter is an example of a custom renderer that performs fast generation of an SVG string for rendering "outside" the DOM.

Next, you may want to take a look at the [experimental Vega WebGL renderer](https://github.com/vega/vega-webgl-renderer). While the WebGL specifics may not matter to you, this repo shows how to "plug-in" a new renderer type that can be invoked from the Vega View component.

Good luck! Though I should probably warn you that the current renderers have been iteratively improved over multiple years of Vega releases. There are a lot of fine details to get right. Even more so if you aim to support both rendering and interaction handling!

**Comment by jheer** [comment]
Closing out the issue, given the answer above.
